Master: two tikv don't report region heartbeat after inject fault to pd leader

## Bug Report

<!-- Thanks for your bug report! Don't worry if you can't fill out all the sections. -->

### What version of TiKV are you using?
<!-- You can run `tikv-server --version` -->
./tikv-server -V
 TiKV
Release Version:   6.2.0-alpha
Edition:           Community
Git Commit Hash:   a9c3e56552c803642f640e9b3fa8725aa3072400
Git Commit Branch: heads/refs/tags/v6.1.0-nightly
UTC Build Time:    2022-06-24 11:02:31
### What operating system and CPU are you using?
<!-- If you're using Linux, you can run `cat /proc/cpuinfo` -->

### Steps to reproduce
<!-- If possible, provide a recipe for reproducing the error. A complete runnable program is good. -->

plan/850383
clinic
    /portal/#/orgs/31/clusters/7113750135811495492?from=1656298923&to=1656335643

[2022/06/27 12:58:56.346 +08:00] [INFO] [chaos.go:93] ["Run chaos"] [name="pd io delay"] [selectors="[endless-oltp-tps-850383-1-635/tc-pd-1]"] [SelectorsRetainPolicy(selectors)="[endless-oltp-tps-850383-1-635/tc-pd-1]"] [targetSelectors="[nil]"] [
TargetSelectorsRetainPolicy(targetSelectors)="[nil]"] [experimentSpec="IODelaySpec{Duration: \"\", Scheduler: <nil>, Delay: \"438ms\", Path: \"\", Percent: 72}"]
[2022/06/27 13:03:56.369 +08:00] [INFO] [chaos.go:156] ["Clean chaos"] [name="pd io delay"] [chaosId="ns=endless-oltp-tps-850383-1-635,kind=io-delay,name=io-delay-xknijuyf,spec=&k8s.ChaosIdentifier{Namespace:\"endless-oltp-tps-850383-1-635\", Name
:\"io-delay-xknijuyf\", Spec:IODelaySpec{Duration: \"\", Scheduler: <nil>, Delay: \"438ms\", Path: \"\", Percent: 72}}"]

### What did you expect?
tikv should works fine

### What did happened?
Two tikvs don't report region heartbeat.
![image](https://user-images.githubusercontent.com/9443637/176410395-deb88135-345e-49ec-a9e1-1fde3e199e9b.png)

```
[2022/06/27 04:59:09.479 +00:00] [ERROR] [client.rs:631] ["failed to send heartbeat"] [err_code=KV:PD:gRPC] [err="Grpc(RpcFinished(Some(RpcStatus { code: 14-UNAVAILABLE, message: \"not leader\", details: [] })))"]
[2022/06/27 04:59:09.800 +00:00] [ERROR] [util.rs:447] ["request failed, retry"] [err_code=KV:PD:gRPC] [err="Grpc(RpcFailure(RpcStatus { code: 14-UNAVAILABLE, message: \"not leader\", details: [] }))"]
```

Found some TiKVs were keeping reconnecting the PD who is not leader.
It could be introduced by https://github.com/tikv/tikv/pull/12629. @Connor1996 Do you have any idea to fix this issue?

> Turns out it was caused by PD RegionHeartbeat RPC returns an "OK" response, then the stream exited on TiKV side without any retrying mechanism. ![yqiirc4368](https://user-images.githubusercontent.com/20279863/177092800-6e470cc1-eee1-4606-b472-600772b39fff.jpg) ![origin_img_v2_daf0f176-a759-4003-b050-7de3dcefbb1g](https://user-images.githubusercontent.com/20279863/177092621-2b81a53a-b0eb-458a-a116-12414966b176.jpg)
> 
> Think we need to fix this issue on PD.

I think the potential problem is this code:
https://github.com/tikv/tikv/blob/dc7c48d1731e079ae0949694f88858554982136e/components/pd_client/src/client.rs#L644-L647.

TIKV will not trigger the event that will reconnect pd  client even if the pd respinse is  the grpc errors.

 one case is that:
Build Link: https://ci.pingcap.net/blue/organizations/jenkins/tiflash_regression_test_daily/detail/tiflash_regression_test_daily/2605/pipeline
Job Page: https://ci.pingcap.net/blue/organizations/jenkins/tiflash_regression_test_daily

```
[2022/07/20 06:41:07.720 +00:00] [ERROR] [client.rs:631] ["failed to send heartbeat"] [err_code=KV:PD:gRPC] [err="Grpc(RpcFinished(Some(RpcStatus { code: 14-UNAVAILABLE, message: \"not leader\", details: [] })))"]

[2022/07/21 17:38:39.343 +08:00] [ERROR] [util.rs:445] ["request failed, retry :unknown error TrySendError { kind: Disconnected }"]

```

--- a/components/error_code/src/pd.rs
+++ b/components/error_code/src/pd.rs
@@ -8,6 +8,7 @@ define_error_codes!(
     CLUSTER_NOT_BOOTSTRAPPED => ("ClusterNotBootstraped", "", ""),
     INCOMPATIBLE => ("Imcompatible", "", ""),
     GRPC => ("gRPC", "", ""),
+    STREAM_DISCONNECT => ("STREAM_DISCONNECT","",""),
     REGION_NOT_FOUND => ("RegionNotFound", "", ""),
     STORE_TOMBSTONE => ("StoreTombstone", "", ""),
     GLOBAL_CONFIG_NOT_FOUND => ("GlobalConfigNotFound","",""),

--- a/components/error_code/src/pd.rs
+++ b/components/error_code/src/pd.rs
@@ -8,7 +8,7 @@ define_error_codes!(
     CLUSTER_NOT_BOOTSTRAPPED => ("ClusterNotBootstraped", "", ""),
     INCOMPATIBLE => ("Imcompatible", "", ""),
     GRPC => ("gRPC", "", ""),
-    STREAM_DISCONNECT => ("STREAM_DISCONNECT","",""),
+    STREAM_DISCONNECT => ("StreamDisconnect","",""),
     REGION_NOT_FOUND => ("RegionNotFound", "", ""),
     STORE_TOMBSTONE => ("StoreTombstone", "", ""),
     GLOBAL_CONFIG_NOT_FOUND => ("GlobalConfigNotFound","",""),

--- a/components/pd_client/src/client.rs
+++ b/components/pd_client/src/client.rs
@@ -619,7 +619,7 @@ impl PdClient for RpcClient {
                             if last > last_report {
                                 last_report = last - 1;
                             }
-                            fail::fail_point!("region_heartbeat_send_failed", |_|{ 
+                            fail::fail_point!("region_heartbeat_send_failed", |_| {
                                 Err(Error::Grpc(grpcio::Error::RemoteStopped))
                             });
                             Ok((r, WriteFlags::default()))
@@ -646,10 +646,8 @@ impl PdClient for RpcClient {
                 .expect("expect region heartbeat sender");
             let ret = sender
                 .unbounded_send(req)
-                .map_err(|e|{
-                    Error::StreamDisconnect(e.into_send_error())
-                });
-                   
+                .map_err(|e| Error::StreamDisconnect(e.into_send_error()));
+
             Box::pin(future::ready(ret)) as PdFuture<_>
         };
 
@@ -1054,8 +1052,7 @@ impl PdClient for RpcClient {
                 .expect("expect region buckets sender");
             let ret = sender
                 .unbounded_send(req)
-                .map_err(|e|
-                    Error::StreamDisconnect(e.into_send_error()));
+                .map_err(|e| Error::StreamDisconnect(e.into_send_error()));
             Box::pin(future::ready(ret)) as PdFuture<_>
         };
 

--- a/components/pd_client/src/client.rs
+++ b/components/pd_client/src/client.rs
@@ -619,6 +619,9 @@ impl PdClient for RpcClient {
                             if last > last_report {
                                 last_report = last - 1;
                             }
+                            fail::fail_point!("region_heartbeat_send_failed", |_|{ 
+                                Err(Error::Grpc(grpcio::Error::RemoteStopped))
+                            });
                             Ok((r, WriteFlags::default()))
                         }))
                         .await;
@@ -643,7 +646,10 @@ impl PdClient for RpcClient {
                 .expect("expect region heartbeat sender");
             let ret = sender
                 .unbounded_send(req)
-                .map_err(|e| Error::Other(Box::new(e)));
+                .map_err(|e|{
+                    Error::SteamDisconnect(e.into_send_error())
+                });
+                   
             Box::pin(future::ready(ret)) as PdFuture<_>
         };
 
@@ -1048,7 +1054,8 @@ impl PdClient for RpcClient {
                 .expect("expect region buckets sender");
             let ret = sender
                 .unbounded_send(req)
-                .map_err(|e| Error::Other(Box::new(e)));
+                .map_err(|e|
+                    Error::SteamDisconnect(e.into_send_error()));
             Box::pin(future::ready(ret)) as PdFuture<_>
         };
 

--- a/components/pd_client/src/client.rs
+++ b/components/pd_client/src/client.rs
@@ -647,7 +647,7 @@ impl PdClient for RpcClient {
             let ret = sender
                 .unbounded_send(req)
                 .map_err(|e|{
-                    Error::SteamDisconnect(e.into_send_error())
+                    Error::StreamDisconnect(e.into_send_error())
                 });
                    
             Box::pin(future::ready(ret)) as PdFuture<_>
@@ -1055,7 +1055,7 @@ impl PdClient for RpcClient {
             let ret = sender
                 .unbounded_send(req)
                 .map_err(|e|
-                    Error::SteamDisconnect(e.into_send_error()));
+                    Error::StreamDisconnect(e.into_send_error()));
             Box::pin(future::ready(ret)) as PdFuture<_>
         };
 

--- a/components/pd_client/src/errors.rs
+++ b/components/pd_client/src/errors.rs
@@ -3,7 +3,7 @@
 use std::{error, result};
 
 use error_code::{self, ErrorCode, ErrorCodeExt};
-use futures::channel::mpsc::{SendError};
+use futures::channel::mpsc::SendError;
 use thiserror::Error;
 
 #[derive(Debug, Error)]
@@ -17,7 +17,7 @@ pub enum Error {
     #[error("{0}")]
     Grpc(#[from] grpcio::Error),
     #[error("{0}")]
-    StreamDisconnect (#[from] SendError),
+    StreamDisconnect(#[from] SendError),
     #[error("unknown error {0:?}")]
     Other(#[from] Box<dyn error::Error + Sync + Send>),
     #[error("region is not found for key {}", log_wrappers::Value::key(.0))]
@@ -33,9 +33,9 @@ pub type Result<T> = result::Result<T, Error>;
 impl Error {
     pub fn retryable(&self) -> bool {
         match self {
-            Error::Grpc(_) | Error::ClusterNotBootstrapped(_)|Error::StreamDisconnect(_) => true,
+            Error::Grpc(_) | Error::ClusterNotBootstrapped(_) | Error::StreamDisconnect(_) => true,
             Error::Other(_)
-            |Error::RegionNotFound(_)
+            | Error::RegionNotFound(_)
             | Error::StoreTombstone(_)
             | Error::GlobalConfigNotFound(_)
             | Error::ClusterBootstrapped(_)

--- a/components/pd_client/src/errors.rs
+++ b/components/pd_client/src/errors.rs
@@ -3,6 +3,7 @@
 use std::{error, result};
 
 use error_code::{self, ErrorCode, ErrorCodeExt};
+use futures::channel::mpsc::{SendError};
 use thiserror::Error;
 
 #[derive(Debug, Error)]
@@ -15,6 +16,8 @@ pub enum Error {
     Incompatible,
     #[error("{0}")]
     Grpc(#[from] grpcio::Error),
+    #[error("{0}")]
+    SteamDisconnect(#[from]SendError),
     #[error("unknown error {0:?}")]
     Other(#[from] Box<dyn error::Error + Sync + Send>),
     #[error("region is not found for key {}", log_wrappers::Value::key(.0))]
@@ -30,9 +33,9 @@ pub type Result<T> = result::Result<T, Error>;
 impl Error {
     pub fn retryable(&self) -> bool {
         match self {
-            Error::Grpc(_) | Error::ClusterNotBootstrapped(_) => true,
+            Error::Grpc(_) | Error::ClusterNotBootstrapped(_)|Error::SteamRpcDisconnect(_) => true,
             Error::Other(_)
-            | Error::RegionNotFound(_)
+            |Error::RegionNotFound(_)
             | Error::StoreTombstone(_)
             | Error::GlobalConfigNotFound(_)
             | Error::ClusterBootstrapped(_)
@@ -48,6 +51,7 @@ impl ErrorCodeExt for Error {
             Error::ClusterNotBootstrapped(_) => error_code::pd::CLUSTER_NOT_BOOTSTRAPPED,
             Error::Incompatible => error_code::pd::INCOMPATIBLE,
             Error::Grpc(_) => error_code::pd::GRPC,
+            Error::SteamDisconnect(_) => error_code::pd::STREAM_DISCONNECT,
             Error::RegionNotFound(_) => error_code::pd::REGION_NOT_FOUND,
             Error::StoreTombstone(_) => error_code::pd::STORE_TOMBSTONE,
             Error::GlobalConfigNotFound(_) => error_code::pd::GLOBAL_CONFIG_NOT_FOUND,

--- a/components/pd_client/src/errors.rs
+++ b/components/pd_client/src/errors.rs
@@ -17,7 +17,7 @@ pub enum Error {
     #[error("{0}")]
     Grpc(#[from] grpcio::Error),
     #[error("{0}")]
-    SteamDisconnect(#[from]SendError),
+    StreamDisconnect (#[from] SendError),
     #[error("unknown error {0:?}")]
     Other(#[from] Box<dyn error::Error + Sync + Send>),
     #[error("region is not found for key {}", log_wrappers::Value::key(.0))]
@@ -33,7 +33,7 @@ pub type Result<T> = result::Result<T, Error>;
 impl Error {
     pub fn retryable(&self) -> bool {
         match self {
-            Error::Grpc(_) | Error::ClusterNotBootstrapped(_)|Error::SteamRpcDisconnect(_) => true,
+            Error::Grpc(_) | Error::ClusterNotBootstrapped(_)|Error::StreamDisconnect(_) => true,
             Error::Other(_)
             |Error::RegionNotFound(_)
             | Error::StoreTombstone(_)
@@ -51,7 +51,7 @@ impl ErrorCodeExt for Error {
             Error::ClusterNotBootstrapped(_) => error_code::pd::CLUSTER_NOT_BOOTSTRAPPED,
             Error::Incompatible => error_code::pd::INCOMPATIBLE,
             Error::Grpc(_) => error_code::pd::GRPC,
-            Error::SteamDisconnect(_) => error_code::pd::STREAM_DISCONNECT,
+            Error::StreamDisconnect(_) => error_code::pd::STREAM_DISCONNECT,
             Error::RegionNotFound(_) => error_code::pd::REGION_NOT_FOUND,
             Error::StoreTombstone(_) => error_code::pd::STORE_TOMBSTONE,
             Error::GlobalConfigNotFound(_) => error_code::pd::GLOBAL_CONFIG_NOT_FOUND,

--- a/components/test_pd/src/mocker/service.rs
+++ b/components/test_pd/src/mocker/service.rs
@@ -238,6 +238,7 @@ impl PdMocker for Service {
             .insert(region_id, req.get_leader().clone());
 
         let mut resp = RegionHeartbeatResponse::default();
+        resp.set_region_id(req.get_region().get_id());
         let header = Service::header();
         resp.set_header(header);
         Some(Ok(resp))

--- a/tests/integrations/pd/test_rpc_client.rs
+++ b/tests/integrations/pd/test_rpc_client.rs
@@ -476,10 +476,10 @@ fn test_change_leader_async() {
 }
 
 #[test]
-fn test_pd_client_heartbeat_send_failed(){
+fn test_pd_client_heartbeat_send_failed() {
     let pd_client_send_fail_fp = "region_heartbeat_send_failed";
     fail::cfg(pd_client_send_fail_fp, "return()").unwrap();
-    let server = MockServer::with_case(1,Arc::new(AlreadyBootstrapped));
+    let server = MockServer::with_case(1, Arc::new(AlreadyBootstrapped));
     let eps = server.bind_addrs();
 
     let client = new_client(eps, None);
@@ -489,38 +489,35 @@ fn test_pd_client_heartbeat_send_failed(){
         .build()
         .unwrap();
     let (tx, rx) = mpsc::channel();
-    let f = client.handle_region_heartbeat_response(1, move |resp| {
-        tx.send(resp).unwrap_or_default()
-    });
+    let f =
+        client.handle_region_heartbeat_response(1, move |resp| tx.send(resp).unwrap_or_default());
     poller.spawn(f);
 
     let heartbeat_send_fail = |ok| {
         let mut region = metapb::Region::default();
         region.set_id(1);
-        let peer = metapb::Peer::default();
-        let stat = RegionStat::default();
         poller.spawn(client.region_heartbeat(
             store::RAFT_INIT_LOG_TERM,
-            region.clone(),
-            peer.clone(),
-            stat.clone(),
+            region,
+            metapb::Peer::default(),
+            RegionStat::default(),
             None,
         ));
-        let rsp=rx.recv_timeout(Duration::from_millis(100));
-        if ok{
+        let rsp = rx.recv_timeout(Duration::from_millis(100));
+        if ok {
             assert!(rsp.is_ok());
-            assert_eq!(rsp.unwrap().get_region_id(),1);
-        }else{
+            assert_eq!(rsp.unwrap().get_region_id(), 1);
+        } else {
             assert!(rsp.is_err());
         }
-       
+
         let region = block_on(client.get_region_by_id(1));
-        if ok{
+        if ok {
             assert!(region.is_ok());
-            let r=region.unwrap();
+            let r = region.unwrap();
             assert!(r.is_some());
-            assert_eq!(1,r.unwrap().get_id());
-        }else{
+            assert_eq!(1, r.unwrap().get_id());
+        } else {
             assert!(region.is_err());
         }
     };

--- a/tests/integrations/pd/test_rpc_client.rs
+++ b/tests/integrations/pd/test_rpc_client.rs
@@ -476,6 +476,62 @@ fn test_change_leader_async() {
 }
 
 #[test]
+fn test_pd_client_heartbeat_send_failed(){
+    let pd_client_send_fail_fp = "region_heartbeat_send_failed";
+    fail::cfg(pd_client_send_fail_fp, "return()").unwrap();
+    let server = MockServer::with_case(1,Arc::new(AlreadyBootstrapped));
+    let eps = server.bind_addrs();
+
+    let client = new_client(eps, None);
+    let poller = Builder::new_multi_thread()
+        .thread_name(thd_name!("poller"))
+        .worker_threads(1)
+        .build()
+        .unwrap();
+    let (tx, rx) = mpsc::channel();
+    let f = client.handle_region_heartbeat_response(1, move |resp| {
+        tx.send(resp).unwrap_or_default()
+    });
+    poller.spawn(f);
+
+    let heartbeat_send_fail = |ok| {
+        let mut region = metapb::Region::default();
+        region.set_id(1);
+        let peer = metapb::Peer::default();
+        let stat = RegionStat::default();
+        poller.spawn(client.region_heartbeat(
+            store::RAFT_INIT_LOG_TERM,
+            region.clone(),
+            peer.clone(),
+            stat.clone(),
+            None,
+        ));
+        let rsp=rx.recv_timeout(Duration::from_millis(100));
+        if ok{
+            assert!(rsp.is_ok());
+            assert_eq!(rsp.unwrap().get_region_id(),1);
+        }else{
+            assert!(rsp.is_err());
+        }
+       
+        let region = block_on(client.get_region_by_id(1));
+        if ok{
+            assert!(region.is_ok());
+            let r=region.unwrap();
+            assert!(r.is_some());
+            assert_eq!(1,r.unwrap().get_id());
+        }else{
+            assert!(region.is_err());
+        }
+    };
+    // send fail if network is block.
+    heartbeat_send_fail(false);
+    fail::remove(pd_client_send_fail_fp);
+    // send success after network recovered.
+    heartbeat_send_fail(true);
+}
+
+#[test]
 fn test_region_heartbeat_on_leader_change() {
     let eps_count = 3;
     let server = MockServer::with_case(eps_count, Arc::new(LeaderChange::new()));

